File: typescript/answers/step3-getter-setter.ts
// Step 3: getter / setter を実装する — 回答例
//
// 実行方法:
//   npm run answer3
//   ※ package.json の scripts に "answer3": "ts-node answers/step3-getter-setter.ts" と定義されているため、
//      npm run answer3 は ts-node answers/step3-getter-setter.ts の短縮呼び出しになる
//
// 末尾の `export {}` について:
//   src/ と answers/ で同じクラス名・変数名が存在するため、モジュール扱いにしてスコープを分離している。
//   実装には影響しないので気にしなくてよい。

class Task {
  public title: string;
  public dueDate: string;

  // private プロパティに _ をつけるのは「直接アクセスしないでください」という慣習
  // getter / setter と名前が衝突しないようにするためでもある
  private _completed: boolean;

  constructor(title: string, dueDate: string) {
    this.title = title;
    this.dueDate = dueDate;
    this._completed = false;
  }

  // getter: _completed を外部に公開する読み取り専用の窓口
  // task.completed と書くだけでアクセスでき、呼び出し側はメソッドと意識しなくてよい
  get completed(): boolean {
    return this._completed;
  }

  // setter: 外部からの書き込みをバリデーションで守る
  // 「完了 → 未完了」への変更を禁止することで、タスクの状態が逆戻りしない保証を持てる
  set completed(value: boolean) {
    // 「すでに完了済み（_completed === true）かつ false に戻そうとしている（!value === true）」場合のみエラー
    if (this._completed && !value) {
      throw new Error("完了済みのタスクを未完了に戻すことはできません");
    }
    this._completed = value;
  }

  // getter はプロパティのように見えるが、呼ばれるたびに計算できる
  // 将来タスクにサブタスクが増えても、この getter だけ修正すれば呼び出し側は変わらない
  get completionRate(): number {
    return this._completed ? 100 : 0;
  }

  display(): void {
    const status = this._completed ? "完了" : "未完了";
    console.log(`[${status}] ${this.title}（期日: ${this.dueDate}）完了率: ${this.completionRate}%`);
  }
}

// --- 動作確認 ---
const task = new Task("買い物", "2024-12-31");
task.display(); // [未完了] 買い物（期日: 2024-12-31）完了率: 0%

task.completed = true;
task.display(); // [完了] 買い物（期日: 2024-12-31）完了率: 100%

// setter のバリデーション確認
try {
  task.completed = false; // 完了済みを未完了に戻そうとするとエラー
} catch (e) {
  console.log((e as Error).message); // 完了済みのタスクを未完了に戻すことはできません
}

export {};

File: python/answers/step3_getter_setter.py
# Step 3: @property / @setter を実装する — 回答例
#
# 実行方法:
#   python answers/step3_getter_setter.py

class Task:
    def __init__(self, title: str, due_date: str) -> None:
        self.title = title
        self.due_date = due_date
        # @property のバッキングストアには慣習的に _ をつける
        # getter / setter と名前が衝突しないようにするためでもある
        self._completed: bool = False

    @property
    def completed(self) -> bool:
        # TypeScript の get completed(): boolean と同等
        # task.completed と書くだけで呼び出せる（メソッドに見えない）
        return self._completed

    @completed.setter
    def completed(self, value: bool) -> None:
        # TypeScript の set completed(value: boolean) と同等
        # setter があることで task.completed = True と書ける
        # setter を定義しなければ読み取り専用になる（TypeScript の readonly に相当）
        #
        # setter でバリデーションを挟むことで不正な状態遷移を防ぐ
        # 「完了 → 未完了」への変更を禁止することで、タスクの状態が逆戻りしない保証を持てる
        if self._completed and not value:
            raise ValueError("完了済みのタスクを未完了に戻すことはできません")
        self._completed = value

    @property
    def completion_rate(self) -> int:
        # setter を定義しないため、外部からは読み取り専用になる
        # TypeScript の readonly getter に相当
        # @property はプロパティのように見えるが、呼ばれるたびに計算できる
        # 将来タスクにサブタスクが増えても、この getter だけ修正すれば呼び出し側は変わらない
        return 100 if self._completed else 0

    def display(self) -> None:
        status = "完了" if self._completed else "未完了"
        print(f"[{status}] {self.title}（期日: {self.due_date}）完了率: {self.completion_rate}%")


# --- 動作確認 ---
task = Task("買い物", "2024-12-31")
task.display()  # [未完了] 買い物（期日: 2024-12-31）完了率: 0%

task.completed = True
task.display()  # [完了] 買い物（期日: 2024-12-31）完了率: 100%

# setter のバリデーション確認
try:
    task.completed = False  # 完了済みを未完了に戻そうとするとエラー
except ValueError as e:
    print(e)  # 完了済みのタスクを未完了に戻すことはできません
